### PR TITLE
petsc: add a +mkl-pardiso variant

### PR DESCRIPTION
**Pull request by lpoirel**
mkl_pardiso solver is distributed with intel-mkl
related to issue #21197

### CODE REVIEW

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 1]
```diff
@@ -92,6 +92,8 @@ class Petsc(Package):
             description='Activates support for SuperluDist (only parallel)')
     variant('trilinos', default=False,
             description='Activates support for Trilinos (only parallel)')
+    variant('mkl-pardiso', default=False,
```
Comment (lpoirel) [inline var/spack/repos/builtin/packages/petsc/package.py:95]: I was not sure what the variant default should be :
maybe we want to build with mkl_pardiso if "intel-mkl" in spec ?

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 2]
```diff
@@ -243,6 +245,7 @@ class Petsc(Package):
     depends_on('trilinos@12.6.2:+mpi', when='@3.7.0:+trilinos+mpi')
     depends_on('trilinos@xsdk-0.2.0+mpi', when='@xsdk-0.2.0+trilinos+mpi')
     depends_on('trilinos@develop+mpi', when='@xdevelop+trilinos+mpi')
+    depends_on('intel-mkl', when='+mkl-pardiso')
```
Comment (lpoirel) [inline var/spack/repos/builtin/packages/petsc/package.py]: it does what we want :
```
$ spack spec petsc +mkl-pardiso ^openblas
Input spec
--------------------------------
petsc+mkl-pardiso
    ^openblas

Concretized
--------------------------------
==> Error: Multiple providers found for 'blas': ['intel-mkl ^cpio', 'openblas']
```

**Comment by lpoirel** [comment]
@balay can you review this pull request?

File: var/spack/repos/builtin/packages/petsc/package.py (hunk 1 above)
Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py:95]: I think `default=False` is a reasonable default

**Review by balay — APPROVED** [review]

File: var/spack/repos/builtin/packages/petsc/package.py (hunk 2 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/petsc/package.py]:
```suggestion
    depends_on('mkl', when='+mkl-pardiso')
```

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 3]
```diff
@@ -449,6 +452,11 @@ def install(self, spec, prefix):
         else:
             options.append('--with-zlib=0')
 
+        if '+mkl-pardiso' in spec:
+            options.append(
+                '--with-mkl_pardiso-dir=%s' % spec['intel-mkl'].prefix
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/petsc/package.py]:
```suggestion
                '--with-mkl_pardiso-dir=%s' % spec['mkl'].prefix
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]
`mkl` is a virtual package provided by both `intel-mkl` and `intel-parallel-studio+mkl`

**Review by adamjstewart — APPROVED** [review]

**Comment by lpoirel** [comment]
Sorry, I was distracted and labelled the PR and commits "mumps: ****" instead of "petsc". I'll correct that tomorrow

**Pull request merged**